<a href="https://colab.research.google.com/github/ryamaguchi0111/gradio/blob/master/gradio_getting_started.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Getting Started
Gradioを実行するためには3つの手順が必要です。  
簡単な例でそれらを実行します。

参考:https://www.gradio.app/getting_started

## 1. Install Gradio from pip

In [ ]:
!pip install gradio

     |████████████████████████████████| 1.4MB 2.7MB/s 
     |████████████████████████████████| 215kB 18.2MB/s 
     |████████████████████████████████| 61kB 6.3MB/s 
     |████████████████████████████████| 962kB 16.3MB/s 
     |████████████████████████████████| 2.7MB 26.6MB/s 


## 2. 以下のコードを実行します。

In [ ]:
import gradio as gr

def greet(name):
    return f'Hello {name} !'

gr.Interface(fn=greet, inputs='text', outputs='text').launch()

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on External URL: https://46533.gradio.app
Interface loading below...


(<gradio.networking.serve_files_in_background.<locals>.HTTPServer at 0x7f801eeee278>,
 'http://127.0.0.1:7860/',
 'https://46533.gradio.app')

## 3.上のインターフェースは、Pythonノートブック内に自動的に表示されます。スクリプトから実行している場合は、http：// localhost： 46533のブラウザに表示されます。

# The Interface
Gradioは、使いやすいインターフェイスです。  
ほとんどすべてのPython関数をラップできます。  
その関数は、単純な計算から事前学習済みモデルまで何でもできます。  

core interface class は3つのパラメータで初期化されます。
- fn : the function to wrap
- inputs : the input component type(s)
- outputs : the output componet type(s)

これらの3つの引数を使用すると、インターフェースをすばやく作成し、それらをlaunch（）できます。　　
UIコンポーネントの外観や動作を変更したい場合


# Customizable Compinents
入力テキストフィールドをカスタマイズしたい場合　　

たとえば、フィールドを大きくしてテキストヒントを設定したいとしますか？文字列のショートカットを使用する代わりに、Textboxの実際の入力クラスを使用する場合、より多くのカスタマイズ性にアクセスできます。  
サポートされているすべてのコンポーネントのリストとそれらをカスタマイズする方法については、[ドキュメント](https://www.gradio.app/docs)をご覧ください。


In [ ]:
import gradio as gr

def greet(name):
  return "Hello " + name + "!"

gr.Interface(
  greet, 
  gr.inputs.Textbox(lines=2, placeholder="Name Here..."),   # placeholoderで指定した文字がテキストヒントとして出てくる
  "text").launch()

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on External URL: https://49779.gradio.app
Interface loading below...


(<gradio.networking.serve_files_in_background.<locals>.HTTPServer at 0x7f802ee2f5f8>,
 'http://127.0.0.1:7861/',
 'https://49779.gradio.app')

# Multiple Inputs and Outputs
複数の入力と出力を備えた、はるかに複雑な関数があるとしましょう。  
以下の例では、文字列、ブール値、数値を受け取り、文字列と数値を返す関数があります。  
入力コンポーネントと出力コンポーネントのリストを渡す方法を見てみましょう。

In [ ]:
import gradio as gr

def greet(name, is_morning, temperature):
  salutation = "Good morning" if is_morning else "Good evening"
  greeting = "%s %s. It is %s degrees today" % (salutation, name, temperature)
  celsius = (temperature - 32) * 5 / 9
  return greeting, round(celsius, 2)

gr.Interface(
  greet,  # fn 
  ["text", "checkbox", gr.inputs.Slider(0, 100)], # inputs
  ["text", "number"]).launch() # outputs

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on External URL: https://47167.gradio.app
Interface loading below...


(<gradio.networking.serve_files_in_background.<locals>.HTTPServer at 0x7f801c8c7240>,
 'http://127.0.0.1:7864/',
 'https://47167.gradio.app')

コンポーネントをリストにラップするだけです。  
さらに、同じ入力型と戻り値型を持つ複数の関数を比較したい場合は、関数のリストを渡してすばやく比較することもできます

# Working with Images
image to image関数を試してみます。  
画像コンポーネントを使用する場合、  
関数は、指定されたサイズの形状（幅、高さ、3）の配列を受け取ります。  
最後の次元はRGB値を表します。   
numpy配列の形式で画像も返します。

In [ ]:
import gradio as gr
import numpy as np

def sepia(img):
  sepia_filter = np.array([[.393, .769, .189],
                           [.349, .686, .168],
                           [.272, .534, .131]])
  sepia_img = img.dot(sepia_filter.T)
  sepia_img /= sepia_img.max()                          
  return sepia_img

gr.Interface(
  sepia, gr.inputs.Image(shape=(300, 300)), "image").launch()

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Running on External URL: https://19896.gradio.app
Interface loading below...


(<gradio.networking.serve_files_in_background.<locals>.HTTPServer at 0x7f801c8d3080>,
 'http://127.0.0.1:7866/',
 'https://19896.gradio.app')

さらに、画像入力インターフェイスには「編集」ボタンが付属しており、トリミング、反転、回転、描画、および画像へのフィルターの適用のためのツールが開きます。この方法で画像を操作すると、モデルの隠れた欠陥が明らかになることがよくあります

# Netx Steps
これで、Gradioを使い始めることができます。  
利用可能なすべてのコンポーネントのリストを確認するには、[ドキュメント](https://www.gradio.app/docs)をご覧ください。   
Gradioを使用した機械学習モデルの例を確認するには、[ML Examples](https://www.gradio.app/ml_examples)のページをご覧ください。   


# Sharing Interfaces Publivkly & Privacy
launch()メソッドで share=True にすると、インターフェースを簡単に共有することができます。  
```python
gr.Interface(classify_image, image, label).launch(share=True)
```

これにより、誰にでも送信できる共有可能な共有リンクが生成されます。このリンクを送信すると、反対側のユーザーはブラウザでモデルを試すことができます。処理はデバイス上で行われるため（デバイスがオンのままである限り）、依存関係を心配する必要はありません。 colabノートブックで作業している場合、共有リンクは常に自動的に作成されます。通常は、**XXXXX.gradio.app**のようになります。リンクはgradioリンクを介して提供されますが、私たちはローカルサーバーのプロキシにすぎず、インターフェースを介して送信されたデータを保存しません。  



ただし、これらのリンクは一般公開されているため、誰でもモデルを予測に使用できます。したがって、作成した関数を通じて機密情報を公開したり、デバイスで重大な変更が発生したりしないようにしてください。 share = False（デフォルト）を設定すると、ローカルリンクのみが作成され、**port-forwarding**によって特定のユーザーと共有できます。  

Links expire after 6 hours. Need longer links, or private links?

